lib/collection/src/collection_manager/segments_updater.rs

```diff
@@ -130,15 +143,43 @@ pub(crate) fn set_payload(
     payload: &Payload,
     points: &[PointIdType],
 ) -> CollectionResult<usize> {
-    let updated_points =
-        segments.apply_points_to_appendable(op_num, points, |id, write_segment| {
-            write_segment.set_payload(op_num, id, payload)
-        })?;
+    let updated_points = segments.apply_points_and_reindex(
+        op_num,
+        points,
+        |id, write_segment| write_segment.set_payload(op_num, id, payload),
+        |segment| safe_to_set_payload(&segment.get_indexed_fields(), payload),
```

Comment:
Note: calling `segment.get_indexed_fields()` is not the fastest way to do it (it builds a whole hashmap, but we only need the keys). But it's only done once per segment and the code is concise.